flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java

```diff
@@ -0,0 +1,409 @@
+            MySqlSnapshotSplit split, SourceRecord highWatermark) {
+        Struct value = (Struct) highWatermark.value();
+        String splitId = value.getString(SPLIT_ID_KEY);
+        return new FinishedSnapshotSplitInfo(
+                split.getTableId(),
+                splitId,
+                split.getSplitStart(),
+                split.getSplitEnd(),
+                getBinlogPosition(highWatermark.sourceOffset()));
+    }
+
+    /** Returns the start offset of the binlog split. */
+    public static BinlogOffset getStartingOffsetOfBinlogSplit(
+            List<FinishedSnapshotSplitInfo> finishedSnapshotSplits) {
+        BinlogOffset startOffset =
+                finishedSnapshotSplits.isEmpty()
+                        ? BinlogOffset.INITIAL_OFFSET
+                        : finishedSnapshotSplits.get(0).getHighWatermark();
+        for (FinishedSnapshotSplitInfo finishedSnapshotSplit : finishedSnapshotSplits) {
+            if (finishedSnapshotSplit.getHighWatermark().isBefore(startOffset)) {
+                startOffset = finishedSnapshotSplit.getHighWatermark();
+            }
+        }
+        return startOffset;
+    }
+
+    public static boolean isDataChangeRecord(SourceRecord record) {
+        Schema valueSchema = record.valueSchema();
+        Struct value = (Struct) record.value();
+        return valueSchema.field(Envelope.FieldName.OPERATION) != null
+                && value.getString(Envelope.FieldName.OPERATION) != null;
+    }
+
+    public static TableId getTableId(SourceRecord dataRecord) {
+        Struct value = (Struct) dataRecord.value();
+        Struct source = value.getStruct(Envelope.FieldName.SOURCE);
+        String dbName = source.getString(DATABASE_NAME_KEY);
+        String tableName = source.getString(TABLE_NAME_KEY);
+        return new TableId(dbName, null, tableName);
+    }
+
+    public static Object[] getSplitKey(
+            RowType splitBoundaryType, SourceRecord dataRecord, SchemaNameAdjuster nameAdjuster) {
+        // the split key field contains single field now
+        String splitFieldName = nameAdjuster.adjust(splitBoundaryType.getFieldNames().get(0));
+        Struct key = (Struct) dataRecord.key();
+        return new Object[] {key.get(splitFieldName)};
+    }
+
+    public static BinlogOffset getBinlogPosition(SourceRecord dataRecord) {
+        return getBinlogPosition(dataRecord.sourceOffset());
+    }
+
+    public static BinlogOffset getBinlogPosition(Map<String, ?> offset) {
+        Map<String, String> offsetStrMap = new HashMap<>();
+        for (Map.Entry<String, ?> entry : offset.entrySet()) {
+            offsetStrMap.put(
+                    entry.getKey(), entry.getValue() == null ? null : entry.getValue().toString());
+        }
+        return new BinlogOffset(offsetStrMap);
+    }
+
+    /** Returns the specific key contains in the split key range or not. */
+    public static boolean splitKeyRangeContains(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.